### PR TITLE
Manually test that panicking from C will abort the process

### LINKED ISSUE #228: Panics through FFI in error callbacks
Right now we override the error callbacks at compile time with:
https://github.com/rust-bitcoin/rust-secp256k1/blob/c6ab14bcbe8050b7fd7cc64a8da262b457ee711c/secp256k1-sys/src/lib.rs#L364-L369
https://github.com/rust-bitcoin/rust-secp256k1/blob/c6ab14bcbe8050b7fd7cc64a8da262b457ee711c/secp256k1-sys/src/lib.rs#L387-L392

But this is sadly UB if the panic strategy is `unwind`(the default):
> Calling a function with the wrong call ABI or unwinding from a function with the wrong unwind ABI.
https://doc.rust-lang.org/reference/behavior-considered-undefined.html

> You must absolutely catch any panics at the FFI boundary
https://doc.rust-lang.org/nomicon/unwinding.html

https://github.com/rust-lang/rust/issues/52652

Now the easiest solution is to replace the panic with abort, but then we have a few problems:
1. We lose the printing, the user just gets a `Aborted (core dumped)` or `Illegal instruction (core dumped)` which makes it very hard to pinpoint (We could conditionally print if `std` is on, but that sucks).
2. Which `abort` do we call?
2.a. The libc one?(adds the assumption of libc which might break for wasm).
2.b. The intrinsic one? (requires nightly).
2.c. `std::process::abort`? requires libstd.
2.d. https://github.com/rust-lang/rfcs/issues/2512

Another potential solutions:
1. Advise users to use `panic=abort` (imo this isn't a good solution).
2. Conditionally use `std::panic::catch_unwind` which will catch the panic if `std` is on but will still print the output and/or stacktrace.
3. Wait for `"C unwind"` which allows unwinding the stack through C. (https://github.com/rust-lang/project-ffi-unwind/blob/master/rfcs/0000-c-unwind-abi.md)
4. Wait for https://github.com/rust-lang/rust/pull/74754/ which allows to condition on if `panic=abort` or not.
5. Dive deep into the machinery of `panic` and see if we can still trigger printing but abort instead, probably won't be possible in stable.
6. Propose new language features to fix this(i.e. `abort!(...)`), but even if they'll get accepted we won't use that release for a long time.

(Thanks to @joshtriplett for some of the suggestions)

### PR DESCRIPTION
**Pull request by real-or-random**
Panicking from C is not UB in newer rust versions and will reliably
trigger an abort (without unwinding). In older rust versions, it is
technically UB but empirically it seems to "just work" (and what should
it realistically do except crashing, which is what we intent).

Since there's potentially no unwinding, we can't test this behavior
using [should_panic]. This PR will check the exit code instead in our
CI tests.

Fixes #228. This is an alternative to #288.

### CODE REVIEW

**Review by apoelstra — DISMISSED** [review]
neat! ack

File: contrib/test.sh [hunk 1]
```diff
@@ -78,6 +78,9 @@ if [ "$DO_ASAN" = true ]; then
     cargo run --release --manifest-path=./no_std_test/Cargo.toml | grep -q "Verified Successfully"
 fi
 
+# Test if panic in C code aborts the process (we need to invert the exit code)
+! cargo test -- --include-ignored --exact 'tests::test_panic_raw_ctx_should_terminate_abnormally'
+
```
Comment (elichai) [inline contrib/test.sh:83]: This currently fails compiling on stable/1.29.0
```
     Running `/home/runner/work/rust-secp256k1/rust-secp256k1/target/debug/deps/secp256k1-a3ae915be309b8a2 --include-ignored --exact 'tests::test_panic_raw_ctx_should_terminate_abnormally'`
error: The "include-ignored" flag is only accepted on the nightly compiler with -Z unstable-options
```

Comment (elichai) [reply contrib/test.sh:83]: This should work on all of them: `cargo +1.29.0 test -- --ignored --exact 'tests::test_panic_raw_ctx_should_terminate_abnormally'`.

Comment (elichai) [reply contrib/test.sh:83]: (I'd suggest we check for SIGILL/SIGABRT exit codes, but on stable/1.29 it will just panic and exit with 101, the same as it exits here for the flag error)

**Review by elichai — CHANGES_REQUESTED** [review]

**Comment by real-or-random** [comment]
Okay fixed... I have no idea about WASM builds but AFAIU this currently does not test if the panic works on WASM.

File: contrib/test.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh -ex
+#!/bin/bash -ex
```
Comment (elichai) [inline contrib/test.sh:1]: I used `sh` instead of `bash` so we could add OSX testing in the future without too much trouble(you probably remember https://github.com/bitcoin-core/secp256k1/pull/750)

Comment (elichai) [inline contrib/test.sh:1]: The downside of `bash` instead of `sh` is that it will make it harder to add OSX testing in the future without too much trouble(you probably remember https://github.com/bitcoin-core/secp256k1/pull/750)

Comment (real-or-random) [reply contrib/test.sh:1]: yeah, let me try again. :D

**Comment by real-or-random** [comment]
Ok this should be really ready for review now...

**Review by elichai — APPROVED** [review]
ACK
Verified that if it panics for another reason it fails
Verified that if it doesn't panic it also fails

**Review by apoelstra — APPROVED** [review]
utack
